lib/pipeline.js

```diff
@@ -852,6 +853,128 @@ class PipelineModel extends BaseModel {
+     *  - [{ "name": "name", "jobs": [1, 2, 3], "pipelineId": 123, "description": "value" },
+     *     { "name": "name2", "jobs": [4, 5], "pipelineId": 123 }]
+     * @method convertStages
+     * @param  {Object}     config              config
+     * @param  {Object}     config.jobs         Jobs
+     * @param  {Object}     config.pipeline     Pipeline
+     * @param  {Object}     config.stages       Pipeline stages
+     * @return {Array}                New array with stages after up-converting
+     */
+    async _convertStages({ pipelineId, stages, jobs, pipelineJobs }) {
+        const newStages = [];
+
+        // Convert stages from object to array of objects
+        Object.entries(stages).forEach(([key, value]) => {
+            const newStage = {
+                name: key,
+                pipelineId,
+                ...value
+            };
+            const jobIds = [];
+
+            // Convert the jobNames to jobIds
+            value.jobs.forEach(jobName => {
+                const job = pipelineJobs.find(j => j.name === jobName);
+
+                if (job) {
+                    jobIds.push(job.id);
+                }
+            });
+
+            newStage.jobIds = jobIds;
+
+            delete newStage.jobs; // extra field from yaml parser
+
+            // Check for setup and teardown
+            const setupJobName = getFullStageName({ stageName: key, type: 'setup' });
+            const teardownJobName = getFullStageName({ stageName: key, type: 'teardown' });
+            const setupJobExists = Object.keys(jobs).find(j => j === setupJobName || j === `~${setupJobName}`);
+            const teardownJobExists = Object.keys(jobs).find(j => j === teardownJobName || j === `~${teardownJobName}`);
+
+            newStage.setup = pipelineJobs.find(j => j.name === setupJobExists).id;
+            newStage.teardown = pipelineJobs.find(j => j.name === teardownJobExists).id;
+
+            newStages.push(newStage);
+        });
+
+        return newStages;
+    }
+
+    /**
+     * Sync stages
+     * 1. Convert new stages into correct format, prepopulate with jobIds
+     * 2.a. Create stages if they are defined and were not already in the database
+     * 2.b. Update existing stages with the new configuration
+     * 2.c. Archive existing stages if they no longer exist in the configuration
+     * @method _createOrUpdateStages
+     * @param  {Object}     config              config
+     * @param  {Object}     config.pipeline     Pipeline
+     * @return {Promise}
+     */
+    async _createOrUpdateStages({ parsedConfig, pipelineId, stageFactory, pipelineJobs }) {
+        // Get new stages
+        const stages = parsedConfig.stages || {};
+        const jobs = parsedConfig.jobs || {};
+
+        // No stages in yaml, skip creation
+        if (Object.keys(stages).length === 0) {
+            logger.info(`No stages for pipelineId:${pipelineId}, skipping stage creation.`);
+
+            return Promise.resolve();
+        }
```

Comment:
If no stages in yaml, we still need to proceed to update (archive) existing stages if any.
```suggestion
```

Reply:
removed